### PR TITLE
Create GitHub release as part of release script and attatch binary distribution

### PR DESCRIPTION
**Pull request by gracegrimwood**
### Type of change

- Enhancement / new feature

### Description

Adds GitHub release creation to release script (with attached binary distribution), updates changelog with recent changes.

More specifically, the script now also runs a maven command to generate a binary distribution zip+tarball, and then runs [a `gh` CLI command](https://cli.github.com/manual/gh_release_create) to create a release (using the contents of `CHANGELOG.md` as the release notes) with the two binary distribution archives attached.

### Additional Context

This means users will be able to download a binary that they can easily extract and run for Kroxylicious, rather than having to download the sources and build these themselves manually. This is required for finishing the quickstart guides ([kroxylicious/kroxylicious.github.io#13](https://github.com/kroxylicious/kroxylicious.github.io/pull/13)).

I have added recent missing changes to `CHANGELOG.md` because we are using this as the release notes for each release.

The problem with doing release notes this way is that it includes _the entire file_ as the notes (previous version sections included), which means in future we're going to end up with some really long release notes. I think we're going to have to revisit this approach at some point, but for now it's the best we've got given how inadequate the GitHub-provided automatically generated notes are (it just gives you a link to the diff between tags). Perhaps at some point we can have another look at generating a list of all merged PRs since the last release for this purpose, but my experiments with it the past few days have revealed how painful it is to format those results into something useable.

### Checklist

_Please go through this checklist and make sure all applicable tasks have been done_

- [ ] Write tests
- [ ] Make sure all tests pass
- [ ] Review performance test results. Ensure that any degradations to performance numbers are understood and justified.
- [ ] Make sure all Sonar-Lint warnings are addressed or are justifiably ignored.
- [ ] Update documentation
- [ ] Reference relevant issue(s) and close them after merging
- [ ] For user facing changes, update CHANGELOG.md (remember to include changes affecting the API of the test artefacts too).

### CODE REVIEW

File: scripts/release.sh
Comment (robobario) [inline scripts/release.sh:148]: does this work if the RELEASE_TAG isn't pushed? I'm guessing `git push "${REPOSITORY}" "${RELEASE_TAG}" ${GIT_DRYRUN:-}` wouldn't have pushed it up.

File: scripts/release.sh
Comment (robobario) [inline scripts/release.sh:145]: maybe we could redundantly do this work setting the default remote higher up and avoid having the duplication between here and lines 172-173

File: scripts/release.sh
Comment (robobario) [inline scripts/release.sh:151]: `--notes-file "CHANGELOG.md"`

does it have some smarts around the notes-file or does it just dump the entire contents as the notes for the release?

Comment (robobario) [inline scripts/release.sh:151]: What do you think of having the title follow the existing pattern of `Release 0.3.0` vs using the release tag like `v0.3.0`?

File: CHANGELOG.md
Comment (SamBarker) [inline CHANGELOG.md:19]: Thanks for catching the missed PRs.

File: scripts/release.sh
Comment (SamBarker) [inline scripts/release.sh:137]: Should we be signing the release artefacts? Should we also be creating md5/sha1 checksums to with the binaries? 

We can do that in a follow up PR if preferred.

File: scripts/release.sh
Comment (SamBarker) [reply scripts/release.sh:151]: Just a dumb copy for now, which I think is fine for this PR. 

Thinking about it maybe we change things so the `changelog.md` represents the current SNAPSHOT release and every time we create a release we create a `changelog-Major.Minor.Patch.md` file?

Comment (SamBarker) [inline scripts/release.sh:151]: `gh release create` is a good start, but I wonder if we should also be creating a PR on the kroxylicious.io git repo to include the new release.

Obviously thats for a separate PR.

**Review by SamBarker — COMMENTED** [review]
I'm going to run a test release before hitting the approve button but it LGTM so far.

File: scripts/release.sh
Comment (gracegrimwood) [reply scripts/release.sh:148]: [From the `gh release create` docs:](https://cli.github.com/manual/gh_release_create)

> If a matching git tag does not yet exist, one will automatically get created from the latest state of the default branch.

So it should be fine if there isn't a release tag already in `$REPOSITORY`. However, there should always be a tag, because the git commands on lines 129-131 handle this before the release is created:
```shell
git tag -f "${RELEASE_TAG}"

git push "${REPOSITORY}" "${RELEASE_TAG}" ${GIT_DRYRUN:-}
```

File: scripts/release.sh
Comment (gracegrimwood) [reply scripts/release.sh:151]: If we were to have `CHANGELOG.md` only contain the changes for the current version (i.e. at the tag for `v0.4.0` it contains all changes between `v0.3.0` and that commit, and on main it contains all changes since the last release) then I think that would achieve what we want. Then if anyone wants to see the changelog for a previous version, they can either go look at the releases page or checkout the commit for that release (which would have a `CHANGELOG.md` containing only that release's changes).

Comment (gracegrimwood) [reply scripts/release.sh:151]: I think it would add more complexity to the script (because I'd have to mess with the `$RELEASE_TAG` string to drop the "v").

Also IMHO it looks better and makes a bit more sense on the GitHub repository homepage if we omit the "Release" bit and stick with just the version number as the title, so it would read

> **Releases**
> v0.3.0

as opposed to the current

> **Releases**
> Release 0.3.0

Having "Release" in there twice feels a bit redundant to me.

File: scripts/release.sh
Comment (gracegrimwood) [reply scripts/release.sh:137]: We probably should, but here I was going for the bare minimum of having a binary artefact attached to the release at all 😉

File: scripts/release.sh
Comment (robobario) [reply scripts/release.sh:151]: works for me, easier to correlate with the tag

File: scripts/release.sh
Comment (SamBarker) [reply scripts/release.sh:137]: Yeah stick something in issues.

File: scripts/release.sh
Comment (k-wall) [reply scripts/release.sh:148]: I'm not against this idea, but I had a thought to experiment with `jreleaser` for  changelog and release note generation. It'd be nice not to have to remember to update the CHANGELOG manually.

https://github.com/ozangunalp/kafka-native/blob/61fe6a3f4b95cd7f404a91d63323efbac5451a02/pom.xml#L409

**Comment by k-wall** [comment]
@gracegrimwood thank you for picking this up.   It'd be good to have the GitHub release automation applied to the junit5 ext too.   Note that the junit5 ext repo already uses automation for the release of the rest of the release process.  There's a pending task to adopt that in this repo too #777

File: scripts/release.sh
Comment (gracegrimwood) [reply scripts/release.sh:148]: @k-wall I agree that updating the changelog manually seems a bit of a hassle, especially if there's an easier way of generating it.

I hadn't heard of `jreleaser`, but it seems like quite a powerful tool that could possibly replace most of the content of `scripts/release.sh`. In this PR, I was just trying to get a binary distribution attached to our releases, so that I can reference that in the quick start guides on the website. Perhaps there's room for another PR where this tool could be properly explored.

File: scripts/release.sh
Comment (gracegrimwood) [reply scripts/release.sh:137]: Agreed that it probably should be in Maven too, but that was beyond the scope of what I was trying to achieve with this PR. Here I was simply trying to ensure we are making a binary distribution available in the repository with each release so it can be referenced in the quick start guide, so that new users don't have to build the whole project from source in order to install it.

**Comment by gracegrimwood** [comment]
@k-wall I saw that issue, but it's a bit beyond the scope of what I was trying to achieve here. In this case, I was just trying to ensure we are adding binary distributions to releases in the repository (which then necessitates automating the creation of GitHub releases), in order to unblock the quick start guide work on the website. I'm happy to come back and revisit #777 in a seperate PR.

**Review by SamBarker — APPROVED** [review]
I have created a trial release using this script which can be found in [my fork](https://github.com/SamBarker/kroxylicious/releases)[1]

The release and thus PR look good. 

n.b. it did stage the jars to maven central but I've dropped that already

[1] I will delete it from my repo once this PR merges.

**Pull request merged**
